dhall/src/Dhall/Pretty/Internal.hs

```diff
@@ -171,107 +171,76 @@ prettySrcExpr = prettyCharacterSet Unicode
-isWhitespace c =
-    case c of
-        ' '  -> True
-        '\n' -> True
-        '\t' -> True
-        '\r' -> True
-        _    -> False
-
-{-| Used to render inline `Src` spans preserved by the syntax tree
-
-    >>> let unusedSourcePos = Text.Megaparsec.SourcePos "" (Text.Megaparsec.mkPos 1) (Text.Megaparsec.mkPos 1)
-    >>> let nonEmptySrc = Src unusedSourcePos unusedSourcePos "-- Documentation for x\n"
-    >>> "let" <> " " <> renderSrc id (Just nonEmptySrc) <> "x = 1 in x"
-    let -- Documentation for x
-        x = 1 in x
-    >>> let emptySrc = Src unusedSourcePos unusedSourcePos "      "
-    >>> "let" <> " " <> renderSrc id (Just emptySrc) <> "x = 1 in x"
-    let x = 1 in x
-    >>> "let" <> " " <> renderSrc id Nothing <> "x = 1 in x"
-    let x = 1 in x
--}
-renderSrc
-    :: (Text -> Text)
-    -- ^ Used to preprocess the comment string (e.g. to strip whitespace)
-    -> Maybe Src
-    -- ^ Source span to render (if present)
-    -> Doc Ann
-renderSrc strip (Just (Src {..}))
-    | not (Text.all isWhitespace srcText) =
-        renderComment (strip srcText)
-renderSrc _ _ =
-    mempty
-
-{-| Render a comment.
-
-    Any preprocessing, such as whitespace stripping, needs to be handled by the
-    caller, see e.g. 'renderSrc'.
-
-    See the documentation for 'renderSrc' for examples.
--}
-renderComment :: Text -> Doc Ann
-renderComment text =
-    Pretty.align (Pretty.concatWith f newLines <> suffix)
+normalizeMultiComment :: MultiComment -> (Bool, [Text])
+normalizeMultiComment (MultiComment comments) =
+    ( any isBlockComment comments
+    , atLeastOneLine . stripEmptyLines $ concatMap toLines comments
+    )
   where
-    horizontalSpace c = c == ' ' || c == '\t'
-
-    suffix =
-        if Text.null text || Text.last text == '\n'
-        then mempty
-        else " "
-
-    oldLines = Text.splitOn "\n" text
-
-    spacePrefix = Text.takeWhile horizontalSpace
-
-    commonPrefix a b = case Text.commonPrefixes a b of
-        Nothing        -> ""
-        Just (c, _, _) -> c
-
-    sharedSpacePrefix []       = ""
-    sharedSpacePrefix (l : ls) = foldl' commonPrefix (spacePrefix l) ls
-
-    blank = Text.all horizontalSpace
-
-    newLines =
-        case oldLines of
-            [] ->
-               []
-            l0 : ls ->
-                let sharedPrefix =
-                        sharedSpacePrefix (filter (not . blank) ls)
-
-                    perLine l =
-                        case Text.stripPrefix sharedPrefix l of
-                            Nothing -> Pretty.pretty l
-                            Just l' -> Pretty.pretty l'
-
-                in  Pretty.pretty l0 : map perLine ls
+    isBlockComment = \case
+        BlockComment _ -> True
+        LineComment _ -> False
+
+    stripStartEmptyLines = dropWhile Text.null
+
+    -- Remove leading and terminating empty lines
+    stripEmptyLines =
+          reverse
+        . stripStartEmptyLines
+        . reverse
+        . stripStartEmptyLines
```

Comment:
You can inline the `dropWhileEnd'` function from here:

https://hackage.haskell.org/package/extra-1.7.9/docs/src/Data.List.Extra.html#dropWhileEnd%27

… which will be more efficient (and lazy) than reversing the list

Reply:
We don't currently depend on `extra` as far as I can tell. Is the version from [`base`](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List.html#v:dropWhileEnd) sufficient?